File: Bai2/Date.cpp
#include "Date.h"
#include <string>
using namespace std;
Date::Date(){
    this->day = 1;
    this->month =1;
    this->year=1900;
}
Date::Date(int d, int m, int y)
{
    if(isDateForm(d,m,y)){
        this->day = d;
        this->month = m;
        this->year = y;
    }else{
        this->day = 1;
        this->month = 1;
        this->year = 1900;
    }
}

string Date::print() {
        return to_string(day) + "/" + to_string(month) + "/" + to_string(year);
    }

   

    bool Date::isLeapYear(int year) 
    {
        return (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0);
    }

    bool Date::isDateForm(int d , int m ,int y) 
    {
        // Kiểm tra tháng hợp lệ
        if (m < 1 || m > 12) {
            return false;
        }
        // Kiểm tra ngày hợp lệ
        if (d < 1) {
            return false;
        }
        switch (m) {
            case 1: case 3: case 5: case 7: case 8: case 10: case 12:
                if (d > 31) {
                    return false;
                }
                break;
            case 4: case 6: case 9: case 11:
                if (d > 30) {
                    return false;
                }
                break;
            case 2:
                if (isLeapYear(y)) {
                    if (d > 29) {
                        return false;
                    }
                } else {
                    if (d > 28) {
                        return false;
                    }
                }
                break;
        }
        return true;
    }

Date::~Date()
{

}

File: Bai2_python/date.py
class Date:
    def __init__(self, day=1, month=1, year=1900):
        if self.is_date_form(day, month, year):
            self.day = day
            self.month = month
            self.year = year
        else:
            self.day = 1
            self.month = 1
            self.year = 1900

    def __str__(self):
        return f"{self.day}/{self.month}/{self.year}"

    @staticmethod
    def is_leap_year(year):
        return (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0)

    @staticmethod
    def is_date_form(d, m, y):
        if m < 1 or m > 12 or d < 1:
            return False
        if m in [1, 3, 5, 7, 8, 10, 12]:
            return d <= 31
        elif m in [4, 6, 9, 11]:
            return d <= 30
        elif m == 2:
            return d <= 29 if Date.is_leap_year(y) else d <= 28
        return True
